### PR TITLE
feat: centralized sac

### PR DESCRIPTION
**Pull request by sash-a**
## What?
Centralized SAC :t-rex:

### CODE REVIEW

File: mava/systems/sac/ff_masac.py [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+@hydra.main(config_path="../../configs", config_name="default_ff_masac.yaml", version_base="1.2")
+def hydra_entry_point(cfg: DictConfig) -> float:
+    """Experiment entry point."""
+    # Allow dynamic attributes.
+    OmegaConf.set_struct(cfg, False)
+
+    # Run experiment.
+    final_return = run_experiment(cfg)
+
+    print(f"{Fore.CYAN}{Style.BRIGHT}ISAC experiment completed{Style.RESET_ALL}")
```
Comment (OmaymaMahjoub) [inline mava/systems/sac/ff_masac.py]: NIT: MASAC :sweat_smile:

File: mava/systems/sac/ff_masac.py [hunk 2]
```diff
@@ -0,0 +1,638 @@
...
+from mava.types import ObservationGlobalState
+from mava.utils import make_env as environments
+from mava.utils.checkpointing import Checkpointer
+from mava.utils.jax import unreplicate_batch_dim, unreplicate_n_dims
+from mava.utils.logger import LogEvent, MavaLogger
+from mava.utils.total_timestep_checker import check_total_timesteps
+from mava.wrappers import episode_metrics
+
+
+def get_joint_action(actions: Array) -> Array:
```
Comment (OmaymaMahjoub) [inline mava/systems/sac/ff_masac.py]: It will be better if we can have this in utils or somewhere else.
Also if you can add docs to mention the motivation behind having this as it is needed for shared action info in the Qs calculation

File: mava/systems/sac/ff_masac.py [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+        log_prob = pi.log_prob(new_action)
+
+        # Repeat the actions from the replay buffer such that you have (B, Ag, Ag, Ac).
+        # This gives you n_agent joint actions with the action dim kept separate.
+        # Then replace along the diagonal with the new action from the policy.
+        # This replacement means that joint_action_i will have the new action for agent_i.
+        actions_repeated = jnp.tile(actions[:, jnp.newaxis, ...], (1, num_agents, 1, 1))
+        inds = jnp.diag_indices_from(actions_repeated[0, ..., 0])
+        spliced_actions = actions_repeated.at[:, inds[0], inds[1], :].set(new_action)
+        joint_actions = spliced_actions.reshape((*spliced_actions.shape[:2], -1))
```
Comment (OmaymaMahjoub) [inline mava/systems/sac/ff_masac.py]: Suggestion: I know we only code this for the actor loss but it will be amazing if we can have it in utils function

**Review by OmaymaMahjoub — CHANGES_REQUESTED** [review]
Thanks @sash-a, just minor suggestions and it will be good to go :fire:

File: mava/systems/sac/ff_masac.py [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+        key: chex.PRNGKey,
+    ) -> Array:
+        pi = actor_net.apply(actor_params, obs)
+        new_action = pi.sample(seed=key)
+        log_prob = pi.log_prob(new_action)
+
+        # Repeat the actions from the replay buffer such that you have (B, Ag, Ag, Ac).
+        # This gives you n_agent joint actions with the action dim kept separate.
+        # Then replace along the diagonal with the new action from the policy.
+        # This replacement means that joint_action_i will have the new action for agent_i.
```
Comment (RuanJohn) [inline mava/systems/sac/ff_masac.py]: Can we please add a legend here for that each axis is referring to? I am not sure what Ag, and Ac are :sweat_smile:

File: mava/utils/centralised_training.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    repeated_action = jnp.tile(actions[:, jnp.newaxis, ...], (1, num_agents, 1, 1))
+    return jnp.reshape(repeated_action, (batch_size, num_agents, act_size * num_agents))
+
+
+def get_updated_joint_actions(rb_actions: Array, policy_actions: Array) -> Array:
+    """
+    Get the updated joint actions by replacing the actions from the replay buffer with the new
+    actions from the policy. Only update joint action i with the new action for agent i.
+
+    The effect of this is that each agents central critic sees what all other agents did in the
```
Comment (RuanJohn) [inline mava/utils/centralised_training.py]:
```suggestion
    The effect of this is that each agent's central critic sees what all other agents did in the
```

File: mava/utils/centralised_training.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    past, but it sees how its agents policy is currently acting.
```
Comment (RuanJohn) [inline mava/utils/centralised_training.py]:
```suggestion
    past, but it sees how its agent's policy is currently acting.
```

File: mava/utils/centralised_training.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+    Method explanation:
+    The `rb_actions` (B, A, Act) will be repeated such that you have two agent dims: (B, A, A, Ac).
```
Comment (RuanJohn) [inline mava/utils/centralised_training.py]:
```suggestion
    The `rb_actions` (B, A, Act) will be repeated such that you have two agent dims: (B, A, A, Act).
```

File: mava/utils/centralised_training.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    Then the diagonal of the repeated actions will be replaced with the new actions from the policy.
+    This replacement means that joint_action[i] will have the new action for agent[i].
+    Finally join the last two dims to get (B, A, A * Act).
```
Comment (RuanJohn) [inline mava/utils/centralised_training.py]:
```suggestion
    Finally join the last two dimensions to get (B, A, A * Act).
```

File: mava/utils/centralised_training.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        rb_actions (B, A, Act): the actions from the replay buffer.
+        policy_actions (B, A, Act): the new actions from the policy.
+
+    Returns: (B, A, A * Act): the updated joint actions.
+    """
+
+    batch_size, num_agents, act_size = rb_actions.shape
+
+    # Repeat the actions from the replay buffer such that you have (B, A, A, Act).
+    # This gives you num agents joint actions with the action dim kept separate.
```
Comment (RuanJohn) [inline mava/utils/centralised_training.py]:
```suggestion
    # This gives num agents joint actions with the action dim kept separate.
```

File: mava/systems/sac/ff_masac.py [hunk 10]
```diff
@@ -0,0 +1,625 @@
...
+        actions: Array,
+        alpha: Array,
+        q_params: QVals,
+        key: chex.PRNGKey,
+    ) -> Array:
+        pi = actor_net.apply(actor_params, obs)
+        new_actions = pi.sample(seed=key)
+        log_prob = pi.log_prob(new_actions)
+
+        # Updated joint actions are done so that each agents central critic sees what all
```
Comment (RuanJohn) [inline mava/systems/sac/ff_masac.py]:
```suggestion
        # Updated joint actions are done so that each agent's central critic sees what all
```

File: mava/systems/sac/ff_masac.py [hunk 11]
```diff
@@ -0,0 +1,625 @@
...
+        # other agents did in the past, but it sees how its agents policy is currently acting.
```
Comment (RuanJohn) [inline mava/systems/sac/ff_masac.py]:
```suggestion
        # other agents did in the past, but it sees how its agent's policy is currently acting.
```

**Review by RuanJohn — CHANGES_REQUESTED** [review]
Last few small suggestions from my side @sash-a. Overall very clean and easy to follow :muscle:

**Review by RuanJohn — APPROVED** [review]
Thank you @sash-a :rocket:

**Review by OmaymaMahjoub — APPROVED** [review]
:fire:
